File: frontend/src/api/samples.ts
// frontend/src/api/samples.ts
import { apiClient } from './client'

// ── Core types ──────────────────────────────────────────────────────────

export interface Sample {
  sample_id: string
  rock_classification: string | null
  state: string | null
  country: string | null
  locality: string | null
  latitude: number | null
  longitude: number | null
  description: string | null
  characterized: boolean
  created_at: string
}

export interface SampleListItem {
  sample_id: string
  rock_classification: string | null
  locality: string | null
  state: string | null
  country: string | null
  characterized: boolean
  experiment_count: number
  has_pxrf: boolean
  has_xrd: boolean
  has_elemental: boolean
  created_at: string
}

export interface SampleListResponse {
  items: SampleListItem[]
  total: number
  skip: number
  limit: number
}

export interface SampleGeoItem {
  sample_id: string
  latitude: number
  longitude: number
  rock_classification: string | null
  characterized: boolean
}

export interface SamplePhoto {
  id: number
  sample_id: string
  file_name: string | null
  file_path: string
  file_type: string | null
  description: string | null
  created_at: string
}

export interface ExternalAnalysis {
  id: number
  sample_id: string | null
  analysis_type: string | null
  analysis_date: string | null
  laboratory: string | null
  analyst: string | null
  pxrf_reading_no: string | null
  description: string | null
  magnetic_susceptibility: string | null
  created_at: string
  analysis_files: AnalysisFile[]
}

export interface AnalysisFile {
  id: number
  external_analysis_id: number
  file_name: string | null
  file_path: string
  file_type: string | null
  created_at: string
}

export interface ActivityEntry {
  id: number
  modification_type: string
  modified_table: string
  modified_by: string
  old_values: Record<string, unknown>
  new_values: Record<string, unknown>
  created_at: string
}

export interface ElementalAnalysisItem {
  analyte_symbol: string
  unit: string
  analyte_composition: number | null
}

export interface LinkedExperiment {
  experiment_id: string
  experiment_type: string | null
  status: string | null
  date: string | null
}

export interface SampleDetail extends Sample {
  photos: SamplePhoto[]
  analyses: ExternalAnalysis[]
  elemental_results: ElementalAnalysisItem[]
  experiments: LinkedExperiment[]
}

export interface ExternalAnalysisCreate {
  analysis_type: string
  analysis_date?: string | null
  laboratory?: string | null
  analyst?: string | null
  pxrf_reading_no?: string | null
  description?: string | null
  magnetic_susceptibility?: string | null
}

export interface AnalysisWithWarnings {
  analysis: ExternalAnalysis
  warnings: string[]
}

// ── List filter params ──────────────────────────────────────────────────

export interface SampleListParams {
  skip?: number
  limit?: number
  country?: string
  rock_classification?: string
  locality?: string
  characterized?: boolean
  search?: string
  has_pxrf?: boolean
  has_xrd?: boolean
  has_elemental?: boolean
}

// ── API client ──────────────────────────────────────────────────────────

export const samplesApi = {
  list: (params?: SampleListParams) =>
    apiClient.get<SampleListResponse>('/samples', { params }).then((r) => r.data),

  listGeo: () =>
    apiClient.get<SampleGeoItem[]>('/samples/geo').then((r) => r.data),

  get: (id: string) =>
    apiClient.get<SampleDetail>(`/samples/${id}`).then((r) => r.data),

  create: (payload: Partial<Sample>) =>
    apiClient.post<Sample>('/samples', payload).then((r) => r.data),

  patch: (id: string, payload: Partial<Sample>) =>
    apiClient.patch<Sample>(`/samples/${id}`, payload).then((r) => r.data),

  delete: (id: string) =>
    apiClient.delete(`/samples/${id}`),

  uploadPhoto: (id: string, file: File, description?: string) => {
    const form = new FormData()
    form.append('file', file)
    if (description) form.append('description', description)
    return apiClient.post<SamplePhoto>(`/samples/${id}/photos`, form).then((r) => r.data)
  },

  deletePhoto: (sampleId: string, photoId: number) =>
    apiClient.delete(`/samples/${sampleId}/photos/${photoId}`),

  listAnalyses: (id: string) =>
    apiClient.get<ExternalAnalysis[]>(`/samples/${id}/analyses`).then((r) => r.data),

  createAnalysis: (id: string, payload: ExternalAnalysisCreate) =>
    apiClient.post<AnalysisWithWarnings>(`/samples/${id}/analyses`, payload).then((r) => r.data),

  deleteAnalysis: (sampleId: string, analysisId: number) =>
    apiClient.delete(`/samples/${sampleId}/analyses/${analysisId}`),

  listActivity: (sampleId: string) =>
    apiClient.get<ActivityEntry[]>(`/samples/${sampleId}/activity`).then((r) => r.data),
}

File: backend/api/schemas/samples.py
# backend/api/schemas/samples.py
from __future__ import annotations
from datetime import date, datetime
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict


# ── Core sample schemas ────────────────────────────────────────────────────

class SampleCreate(BaseModel):
    sample_id: str
    rock_classification: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    locality: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    well_name: Optional[str] = None
    core_lender: Optional[str] = None
    core_interval_ft: Optional[str] = None
    on_loan_return_date: Optional[date] = None


class SampleUpdate(BaseModel):
    rock_classification: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    locality: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    characterized: Optional[bool] = None
    well_name: Optional[str] = None
    core_lender: Optional[str] = None
    core_interval_ft: Optional[str] = None
    on_loan_return_date: Optional[date] = None


class SampleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sample_id: str
    rock_classification: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    locality: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    characterized: bool
    well_name: Optional[str] = None
    core_lender: Optional[str] = None
    core_interval_ft: Optional[str] = None
    on_loan_return_date: Optional[date] = None
    created_at: datetime


# ── List view (no nested objects) ─────────────────────────────────────────

class SampleListItem(BaseModel):
    """Flat projection for the inventory table — no nested objects."""
    model_config = ConfigDict(from_attributes=True)

    sample_id: str
    rock_classification: Optional[str] = None
    locality: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    description: Optional[str] = None
    characterized: bool
    experiment_count: int = 0
    has_pxrf: bool = False
    has_xrd: bool = False
    has_elemental: bool = False
    created_at: datetime


class SampleListResponse(BaseModel):
    items: list[SampleListItem]
    total: int
    skip: int
    limit: int


# ── Geo view (map markers) ────────────────────────────────────────────────

class SampleGeoItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sample_id: str
    latitude: float
    longitude: float
    rock_classification: Optional[str] = None
    characterized: bool


# ── Photo schemas ─────────────────────────────────────────────────────────

class SamplePhotoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sample_id: str
    file_name: Optional[str] = None
    file_path: str
    file_type: Optional[str] = None
    description: Optional[str] = None
    created_at: datetime


# ── Analysis schemas ──────────────────────────────────────────────────────

ANALYSIS_TYPE = Literal["pXRF", "XRD", "Elemental", "Titration", "Magnetic Susceptibility", "Other"]

class ExternalAnalysisCreate(BaseModel):
    analysis_type: ANALYSIS_TYPE  # validated against known types
    analysis_date: Optional[datetime] = None
    laboratory: Optional[str] = None
    analyst: Optional[str] = None
    pxrf_reading_no: Optional[str] = None  # comma-separated reading numbers
    description: Optional[str] = None
    magnetic_susceptibility: Optional[str] = None


class AnalysisFileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    external_analysis_id: int
    file_name: Optional[str] = None
    file_path: str
    file_type: Optional[str] = None
    created_at: datetime


class PXRFElementalData(BaseModel):
    """Averaged elemental values from one or more pXRF readings."""
    reading_count: int
    fe: Optional[float] = None
    mg: Optional[float] = None
    ni: Optional[float] = None
    cu: Optional[float] = None
    si: Optional[float] = None
    co: Optional[float] = None
    mo: Optional[float] = None
    al: Optional[float] = None
    ca: Optional[float] = None
    k: Optional[float] = None
    au: Optional[float] = None
    zn: Optional[float] = None


class XRDPhaseData(BaseModel):
    """Mineral phase data from an XRD analysis."""
    mineral_phases: dict[str, float]
    analysis_parameters: Optional[dict] = None


class ExternalAnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sample_id: Optional[str] = None
    analysis_type: Optional[str] = None
    analysis_date: Optional[datetime] = None
    laboratory: Optional[str] = None
    analyst: Optional[str] = None
    pxrf_reading_no: Optional[str] = None
    description: Optional[str] = None
    magnetic_susceptibility: Optional[str] = None
    created_at: datetime
    analysis_files: list[AnalysisFileResponse] = []
    pxrf_data: Optional[PXRFElementalData] = None
    xrd_data: Optional[XRDPhaseData] = None


class ExternalAnalysisWithWarnings(BaseModel):
    analysis: ExternalAnalysisResponse
    warnings: list[str] = []


# ── Detail view (full nested) ─────────────────────────────────────────────

class LinkedExperiment(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    experiment_id: str
    experiment_type: Optional[str] = None  # from ExperimentalConditions.experiment_type
    status: Optional[str] = None
    date: Optional[datetime] = None


class ElementalAnalysisItem(BaseModel):
    """One analyte row from ElementalAnalysis, for the Analyses tab elemental group."""
    model_config = ConfigDict(from_attributes=True)

    analyte_symbol: str
    unit: str
    analyte_composition: Optional[float] = None


class SampleDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    sample_id: str
    rock_classification: Optional[str] = None
    state: Optional[str] = None
    country: Optional[str] = None
    locality: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    characterized: bool
    well_name: Optional[str] = None
    core_lender: Optional[str] = None
    core_interval_ft: Optional[str] = None
    on_loan_return_date: Optional[date] = None
    created_at: datetime
    photos: list[SamplePhotoResponse] = []
    analyses: list[ExternalAnalysisResponse] = []
    elemental_results: list[ElementalAnalysisItem] = []
    experiments: list[LinkedExperiment] = []
